# Analysis of Hacker News (HN) 
## 2nd Guided Project
In this project, I'll work with a data set of submissions to popular technology site Hacker News. This is m y second guided project on DataQuest, but it is my third in real life.

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. 

Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitted

In this project,I want to determine if certain titles receive more comments on an average and if posts created at a certain time  receive more comments on an average

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
hn_header = hn[0]
hn = hn[1:]

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
print(hn_header)
print('\n')
explore_data(hn, 0, 4, True)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


Number of rows: 20100
Number of columns: 7


In [3]:
explore_data(hn, 0, 5, True)

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']


Number of rows: 20100
Number of columns: 7


In [4]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [5]:
print("The First Five Rows of ask_posts:")
print(hn_header)
print('\n')
print(ask_posts[0:4])
print('\n')
print("The First Five Rows of show_posts:")
print(show_posts[0:4])

The First Five Rows of ask_posts:
['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']]


The First Five Rows of show_posts:
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io'

In [6]:
lst = []
# This is the list that shows all num ber of comments for ask_posts
for row in ask_posts:
    num = int(row[4])
    lst.append(num)
print("Ask_post Comment Summary :")
print("Number of comment enteries:", len(lst))
print("Sum of Comments:", sum(lst))
avg_ask_comments = sum(lst) / len(ask_posts)
print("Average Number of Comments:", avg_ask_comments)

Ask_post Comment Summary :
Number of comment enteries: 1744
Sum of Comments: 24483
Average Number of Comments: 14.038417431192661


In [7]:
slst = []
# This is the list that shows all num ber of comments for show_posts
for row in show_posts:
    num = int(row[4])
    slst.append(num)
print("Show_post Comment Summary:")
print("Number of comment enteries:", len(slst))
print("Sum of Comments:", sum(slst))
avg_show_comments = sum(slst) / len(show_posts)
print("Average Number of Comments:", avg_show_comments)

Show_post Comment Summary:
Number of comment enteries: 1162
Sum of Comments: 11988
Average Number of Comments: 10.31669535283993


## So Far.....
I wrote a function in code cell 2 to explore the dataset. Next, I also explored the entries that matter; `ask_posts` and `show_posts`. In code cell 6 and 7, I tried to get a summary on the Comments for my concerned enteries.
### Ask Posts:
* All posts had comments (1744)
* Sum of Comments: 24483
* Average Number of Comments: 10.31669535283993

### Show Posts:
* All posts had comments (1162)
* Sum of Comments: 11988
* Average Number of Comments: 14.038417431192661


# Next...
From my work above, on average, ask posts received more comments than show posts. Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

* Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
* Calculate the average number of comments ask posts receive by hour created.


## Amount of `ask_posts` Created in Each Hour of the Day, Along with the Number of Comments Received.

In [8]:
import datetime as dt

In [9]:
result_list = []
for row in ask_posts:
    created_at = row[6]
    num = int(row[4])
    
    result_list.append([created_at, num])
print(result_list[:4])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3]]


In [10]:
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = row[0]
    comments = row[1]
    dt_object = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour = dt_object.hour
        
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments
            
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments
    

In [11]:
print("Number of ask_posts created by Hour:")
print(counts_by_hour)
print('\n')
print("Comments by Hour:")
print(comments_by_hour)

Number of ask_posts created by Hour:
{0: 55, 1: 60, 2: 58, 3: 54, 4: 47, 5: 46, 6: 44, 7: 34, 8: 48, 9: 45, 10: 59, 11: 58, 12: 73, 13: 85, 14: 107, 15: 116, 16: 108, 17: 100, 18: 109, 19: 110, 20: 80, 21: 109, 22: 71, 23: 68}


Comments by Hour:
{0: 447, 1: 683, 2: 1381, 3: 421, 4: 337, 5: 464, 6: 397, 7: 267, 8: 492, 9: 251, 10: 793, 11: 641, 12: 687, 13: 1253, 14: 1416, 15: 4477, 16: 1814, 17: 1146, 18: 1439, 19: 1188, 20: 1722, 21: 1745, 22: 479, 23: 543}


## Average Number of Comments `ask_posts` Received by Hour Created

In [12]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, round((comments_by_hour[hour] / counts_by_hour[hour]), 2)])
    
print("Average comments per hour on Ask Hn posts:", '\n', avg_by_hour)

Average comments per hour on Ask Hn posts: 
 [[0, 8.13], [1, 11.38], [2, 23.81], [3, 7.8], [4, 7.17], [5, 10.09], [6, 9.02], [7, 7.85], [8, 10.25], [9, 5.58], [10, 13.44], [11, 11.05], [12, 9.41], [13, 14.74], [14, 13.23], [15, 38.59], [16, 16.8], [17, 11.46], [18, 13.2], [19, 10.8], [20, 21.52], [21, 16.01], [22, 6.75], [23, 7.99]]


In [13]:
swap_avg_by_hour = []

for row in avg_by_hour:
    hour = str(row[0])
    avg_comment = row[1]
    swap_avg_by_hour.append([avg_comment, hour])
    
print(swap_avg_by_hour)



[[8.13, '0'], [11.38, '1'], [23.81, '2'], [7.8, '3'], [7.17, '4'], [10.09, '5'], [9.02, '6'], [7.85, '7'], [10.25, '8'], [5.58, '9'], [13.44, '10'], [11.05, '11'], [9.41, '12'], [14.74, '13'], [13.23, '14'], [38.59, '15'], [16.8, '16'], [11.46, '17'], [13.2, '18'], [10.8, '19'], [21.52, '20'], [16.01, '21'], [6.75, '22'], [7.99, '23']]


In [14]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print(sorted_swap)
print('\n')
print("Top 5 Hours for Ask Posts Comments:")

for row in sorted_swap[:5]:
    top_hour = dt.datetime.strptime(row[1], "%H")
    hour_fmt = top_hour.strftime("%H:%M")
    print("{a} : {b} average comments per post ".format(a=hour_fmt,b=row[0]))



[[38.59, '15'], [23.81, '2'], [21.52, '20'], [16.8, '16'], [16.01, '21'], [14.74, '13'], [13.44, '10'], [13.23, '14'], [13.2, '18'], [11.46, '17'], [11.38, '1'], [11.05, '11'], [10.8, '19'], [10.25, '8'], [10.09, '5'], [9.41, '12'], [9.02, '6'], [8.13, '0'], [7.99, '23'], [7.85, '7'], [7.8, '3'], [7.17, '4'], [6.75, '22'], [5.58, '9']]


Top 5 Hours for Ask Posts Comments:
15:00 : 38.59 average comments per post 
02:00 : 23.81 average comments per post 
20:00 : 21.52 average comments per post 
16:00 : 16.8 average comments per post 
21:00 : 16.01 average comments per post 


### Time Zone: Converting to West-African Time Standard by adding 5 hours

In [15]:
print("Top 5 Hours for Ask Posts Comments(WAT, Nigeria):")
hour_format = "%H"
for row in sorted_swap[:5]:
    hour = row[1]
    # The original time zone was Eastern Standard.
    # Here I am converting to West-African Time Standard by adding 5 hours.
    convert_to_wat = dt.datetime.strptime(hour, hour_format)
    wat = convert_to_wat + dt.timedelta(hours=5)
    wat = wat.strftime("%H:%M")     
   
    avg_comments = round(row[0],2)
    print("{0} {1} average comments per post".format(wat, row[0]))

Top 5 Hours for Ask Posts Comments(WAT, Nigeria):
20:00 38.59 average comments per post
07:00 23.81 average comments per post
01:00 21.52 average comments per post
21:00 16.8 average comments per post
02:00 16.01 average comments per post


# My Conclusions and Findings
From the analysis, the hour that receives the most comments is 15:00 hrs EST (20:00hrs WAT). There are many factors that could affect this, but whichever it is, it is important that the platform owners maximize the hour 